coderd/rbac/scopes.go

```diff
@@ -8,39 +8,52 @@ import (
 
 type Scope string
 
+// TODO: @emyrk rename this struct
+type ScopeRole struct {
+	Role
+	AllowIDList []string `json:"allow_list"`
+}
+
 const (
 	ScopeAll                Scope = "all"
 	ScopeApplicationConnect Scope = "application_connect"
 )
 
-var builtinScopes map[Scope]Role = map[Scope]Role{
+// TODO: Support passing in scopeID list for allowlisting resources.
+var builtinScopes = map[Scope]ScopeRole{
 	// ScopeAll is a special scope that allows access to all resources. During
 	// authorize checks it is usually not used directly and skips scope checks.
 	ScopeAll: {
-		Name:        fmt.Sprintf("Scope_%s", ScopeAll),
-		DisplayName: "All operations",
-		Site: permissions(map[string][]Action{
-			ResourceWildcard.Type: {WildcardSymbol},
-		}),
-		Org:  map[string][]Permission{},
-		User: []Permission{},
+		Role: Role{
+			Name:        fmt.Sprintf("Scope_%s", ScopeAll),
+			DisplayName: "All operations",
+			Site: permissions(map[string][]Action{
+				ResourceWildcard.Type: {WildcardSymbol},
+			}),
+			Org:  map[string][]Permission{},
+			User: []Permission{},
+		},
+		AllowIDList: []string{WildcardSymbol},
 	},
 
 	ScopeApplicationConnect: {
-		Name:        fmt.Sprintf("Scope_%s", ScopeApplicationConnect),
-		DisplayName: "Ability to connect to applications",
-		Site: permissions(map[string][]Action{
-			ResourceWorkspaceApplicationConnect.Type: {ActionCreate},
-		}),
-		Org:  map[string][]Permission{},
-		User: []Permission{},
+		Role: Role{
+			Name:        fmt.Sprintf("Scope_%s", ScopeApplicationConnect),
+			DisplayName: "Ability to connect to applications",
+			Site: permissions(map[string][]Action{
+				ResourceWorkspaceApplicationConnect.Type: {ActionCreate},
+			}),
+			Org:  map[string][]Permission{},
+			User: []Permission{},
+		},
+		AllowIDList: []string{WildcardSymbol},
 	},
 }
 
-func ScopeRole(scope Scope) (Role, error) {
+func ExpandScope(scope Scope) (ScopeRole, error) {
```

Comment:
I'm not sure it makes sense to keep Scope as a string in the database once we are going to have things like an allowlist into it.  Do we need to expand it to a JSON object?

We're not going to be able to keep a map of all workspace scopes in memory.

Reply:
We might need to, but we are going to create the workspace agent authorization context in memory, not from the db.

Agent tokens right now do not have an RBAC subject related to it, it's not an APIKey. So when we implement agent tokens, we will use the workspace owner's roles and add the agent scope to it. Meaning the scope is not in the database.


But you are correct if we use this for more and more cases, it might require us to elevate this to the DB in the future.

Reply:
Yes.

https://github.com/coder/coder/blob/b9fed910857b6ffdc0b6ab5f17fa08f6db491794/coderd/database/dump.sql#L437-L438